![Logo de Icesi](https://www.icesi.edu.co/launiversidad/images/La_universidad/logosimbolos/Logo_icesi_JPG.jpg)

# PROYECTO FINAL DE ANALÍTICA DE DATOS

**Elaborado por:** Santiago Restrepo Prado & Andres Zapata

## Contenido
**1. Limpieza y EDA**

**2. Modelos predictivos**

**3. Caracterización de los clientes**

# Limpieza y EDA

In [0]:
import numpy as np #operaciones matriciales y con vectores
import pandas as pd #tratamiento de datos
import matplotlib.pyplot as plt #gráficos
from sklearn.model_selection import train_test_split #metodo de particionamiento de datasets para evaluación
from sklearn.model_selection import KFold, cross_val_score #protocolo de evaluación
from sklearn import neighbors, datasets, metrics
from sklearn import preprocessing
import seaborn as sns

Importamos los registros del archivo .tsv (separado por tabulador) a un dataframe y lo mostramos

In [2]:
df = pd.read_csv(filepath_or_buffer='./training.tsv', sep="\t", header=0)
df.head(10)

FileNotFoundError: ignored

In [0]:
df.shape

(40053, 232)

El dataset está conformado por **40.053 registros con 232 variables**

In [0]:
df.dtypes

Var1          float64
Var2          float64
Var3          float64
Var4          float64
Var5          float64
Var6          float64
Var7          float64
Var8          float64
Var9          float64
Var10         float64
Var11         float64
Var12         float64
Var13         float64
Var14         float64
Var15         float64
Var16         float64
Var17         float64
Var18         float64
Var19         float64
Var20         float64
Var21         float64
Var22         float64
Var23         float64
Var24         float64
Var25         float64
Var26         float64
Var27         float64
Var28         float64
Var29         float64
Var30         float64
               ...   
Var203         object
Var204         object
Var205         object
Var206         object
Var207         object
Var208         object
Var209        float64
Var210         object
Var211         object
Var212         object
Var213         object
Var214         object
Var215         object
Var216         object
Var217    

In [0]:
df.describe(include='all').transpose()

count unique            top   freq        mean       std    min  \
Var1          549    NaN            NaN    NaN     12.1676    45.251      0   
Var2          987    NaN            NaN    NaN  0.00506586  0.159152      0   
Var3          986    NaN            NaN    NaN     473.182   4738.65      0   
Var4         1242    NaN            NaN    NaN    0.152174   1.41428      0   
Var5         1206    NaN            NaN    NaN      235203    633929      0   
Var6        35641    NaN            NaN    NaN     1322.62   2620.59      0   
Var7        35649    NaN            NaN    NaN     6.80305   6.33762      0   
Var8            0    NaN            NaN    NaN         NaN       NaN    NaN   
Var9          549    NaN            NaN    NaN     50.5392   171.858      0   
Var10        1206    NaN            NaN    NaN      400235    957252      0   
Var11         986    NaN            NaN    NaN     8.64909   3.01074      8   
Var12         443    NaN            NaN    NaN     16.4876   70.6555      0   
Var13       35649    NaN            NaN    NaN     1243.55   2789.89      0   
Var14         986    NaN            NaN    NaN    0.770791   3.88335      0   
Var15           0    NaN            NaN    NaN         NaN       NaN    NaN   
Var16        1206    NaN            NaN    NaN     119.703   71.8966      0   
Var17        1242    NaN            NaN    NaN     9.54509   22.3448      0   
Var18        1242    NaN            NaN    NaN     6.72947   32.6547      0   
Var19        1242    NaN            NaN    NaN    0.188406   1.43287      0   
Var20           0    NaN            NaN    NaN         NaN       NaN    NaN   
Var21       35641    NaN            NaN    NaN     235.027   569.993      0   
Var22       36069    NaN            NaN    NaN     290.822   710.141      0   
Var23        1206    NaN            NaN    NaN     7.86484   53.4167      0   
Var24       34279    NaN            NaN    NaN     4.49978    9.8463      0   
Var25       36069    NaN            NaN    NaN     96.9712   218.408      0   
Var26        1206    NaN            NaN    NaN   0.0771144  0.561436      0   
Var27        1206    NaN            NaN    NaN   0.0281924   0.24955      0   
Var28       36067    NaN            NaN    NaN     224.477   99.4572 -66.88   
Var29         549    NaN            NaN    NaN   0.0255009  0.224596      0   
Var30         549    NaN            NaN    NaN     7.38616   8.55149      0   
...           ...    ...            ...    ...         ...       ...    ...   
Var203      39933      5           9_Y1  36236         NaN       NaN    NaN   
Var204      40053    100           RVjC   1427         NaN       NaN    NaN   
Var205      38495      3           VpdQ  25568         NaN       NaN    NaN   
Var206      35641     21           IYzP  13792         NaN       NaN    NaN   
Var207      40053     14     me75fM6ugJ  28117         NaN       NaN    NaN   
Var208      39933      2           kIsH  36860         NaN       NaN    NaN   
Var209          0    NaN            NaN    NaN         NaN       NaN    NaN   
Var210      40053      6           uKAI  38105         NaN       NaN    NaN   
Var211      40053      2           L84s  32324         NaN       NaN    NaN   
Var212      40053     79        NhsEn4L  23463         NaN       NaN    NaN   
Var213        909      1           KdSa    909         NaN       NaN    NaN   
Var214      19713  13323        5zARyjR     57         NaN       NaN    NaN   
Var215        542      1           eGzu    542         NaN       NaN    NaN   
Var216      40053   1817        mAjbk_S   4021         NaN       NaN    NaN   
Var217      39495  12433           gvA6    222         NaN       NaN    NaN   
Var218      39495      2           cJvF  20319         NaN       NaN    NaN   
Var219      35885     22           FzaX  32303         NaN       NaN    NaN   
Var220      40053   3859        4UxGlow   3574         NaN       NaN    NaN   
Var221      40053      7           oslk  29678         NaN    

De las descripciones obtenidas del dataframe, no parece haber problemas de tipos de datos. Sin embargo, hay columnas que se pueden eliminar porque su columna *count* de la descripción el valor es menor al 50% del total de los datos.

In [0]:
for column in df.columns:
    if (df[column].aggregate('count') < 20027): #20.027 es el 50% de 40053
        df.drop(labels=column, axis=1, inplace=True)
df.shape[1]

71

El resultado obtenido quiere decir que se eliminaron 232 - 71 = 161 columnas

In [0]:
##dfRows = pd.DataFrame()
##for column in df.columns:
##    rows = df[df[column].isna()]
##    dfRows = dfRows.append(rows)
##dfRows

In [0]:
df1 = df[df.isna().any(axis=1)]
df1

Var6  Var7    Var13   Var21   Var22  Var24  Var25   Var28  Var35  \
0      1526.0   7.0    184.0   464.0   580.0   14.0  128.0  166.56    0.0   
1       525.0   0.0      0.0   168.0   210.0    2.0   24.0  353.52    0.0   
2      5236.0   7.0    904.0  1212.0  1515.0   26.0  816.0  220.08    0.0   
3         NaN   0.0      0.0     NaN     0.0    NaN    0.0   22.08    0.0   
5       658.0   7.0   3156.0   224.0   280.0    2.0   72.0  200.00    5.0   
6        77.0   0.0      0.0    32.0    40.0    2.0   16.0  230.56    0.0   
7      1176.0   7.0   2912.0   200.0   250.0    2.0   64.0  300.32    0.0   
8      1141.0   7.0    164.0   208.0   260.0    2.0   72.0  166.56    5.0   
9       490.0   7.0   2060.0    92.0   115.0    6.0  112.0  133.12    0.0   
11      595.0   0.0   2044.0     0.0     0.0    NaN    0.0  240.56    0.0   
12     2268.0   0.0      0.0   480.0   600.0   10.0  216.0  176.56    0.0   
14      259.0   0.0    348.0     8.0    10.0    0.0    0.0  166.56    0.0   
15     1449.0   7.0     28.0   168.0   210.0    2.0   32.0  166.56    5.0   
16      574.0   7.0   1752.0    12.0    15.0    2.0    0.0  253.52    0.0   
18       14.0   0.0      0.0    20.0    25.0    2.0    0.0   86.96    0.0   
19      826.0   0.0     24.0   192.0   240.0    2.0   80.0  166.56    0.0   
20     1043.0   7.0    652.0   152.0   190.0    4.0   16.0   20.08   10.0   
21     1533.0   7.0     84.0   332.0   415.0    6.0  208.0  133.12    0.0   
22      924.0  28.0   4896.0   112.0   140.0    4.0   56.0  166.56    0.0   
23      469.0   7.0   4688.0    28.0    35.0    0.0   16.0  285.20    0.0   
24      833.0  14.0   7124.0   132.0   165.0    2.0   24.0  220.08    0.0   
27        NaN   0.0      0.0     NaN     0.0    NaN    0.0   89.84    0.0   
28     1813.0   0.0      0.0   612.0   765.0   14.0  360.0  200.00    0.0   
31      714.0   7.0   2240.0   120.0   150.0   10.0   80.0  133.12    0.0   
32        NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
34      560.0   0.0      0.0     0.0     0.0    NaN    0.0  120.40    5.0   
35        0.0   0.0      0.0    16.0    20.0    0.0   24.0  397.44    0.0   
37      224.0   0.0      0.0     0.0     0.0    NaN    0.0  186.64    0.0   
38        0.0   0.0      0.0     0.0     0.0    NaN    0.0  274.16    0.0   
39      273.0   0.0      0.0    40.0    50.0    0.0    0.0  220.08    5.0   
...       ...   ...      ...     ...     ...    ...    ...     ...    ...   
39998   105.0   0.0     12.0     4.0     5.0    0.0    0.0  100.32    0.0   
40002  3570.0   7.0    512.0  3268.0  4085.0   58.0  504.0  220.08    0.0   
40004  1792.0   7.0      0.0   160.0   200.0    2.0   48.0  510.96   15.0   
40007     NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
40009  1561.0  21.0    868.0   420.0   525.0   10.0  328.0  133.12    0.0   
40015   196.0   0.0      0.0   160.0   200.0    4.0   72.0  280.56    0.0   
40016  1575.0   7.0    352.0   512.0   640.0    4.0  232.0  166.56    0.0   
40017     NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
40018  5726.0  14.0  12524.0   176.0   220.0    4.0   80.0  240.16    0.0   
40019     NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
40021  1337.0   7.0   1056.0   140.0   175.0    4.0   64.0  253.52    0.0   
40022   546.0   7.0    192.0   192.0   240.0    0.0   72.0  166.56    0.0   
40023  1659.0   0.0      0.0    16.0    20.0    0.0    0.0  220.08    0.0   
40024   133.0   7.0     76.0    36.0    45.0    0.0   16.0  200.00    0.0   
40025     NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
40026   742.0   7.0   2448.0   124.0   155.0    0.0   16.0  220.08   10.0   
40027  1218.0   7.0   3416.0   188.0   235.0    4.0   64.0  233.44    0.0   
40028   602.0   7.0   4952.0   176.0   220.0    4.0  128.0  238.80    0.0   
40030     NaN   NaN      NaN     NaN     NaN    NaN    NaN     NaN    NaN   
40034   511.0   7.0    104.0   116.0   145.0    4.0   96.0  167.20    0.0   
40035

In [0]:
df1.to_csv(path_or_buf='./rows.tsv', sep="\t")

In [0]:
df2 = df1.isna().sum(axis=1)
df3 = df2.sort_values(axis=0, ascending=True)
df3

27685     1
19939     1
36005     1
19945     1
36009     1
8796      1
8795      1
36011     1
19968     1
8785      1
36022     1
8779      1
8776      1
8773      1
31021     1
8768      1
19975     1
8765      1
19977     1
19979     1
19984     1
36024     1
36025     1
19996     1
8747      1
31017     1
8812      1
36003     1
19931     1
19929     1
         ..
32744    42
7634     42
34332    42
18311    42
7616     42
16258    42
37420    42
10096    42
17662    42
25231    42
19636    42
38382    42
16693    42
12095    42
8760     42
18461    42
38545    42
27464    42
2352     42
572      42
39368    42
37135    42
3986     42
29520    43
29562    43
20980    43
9441     43
26293    43
17801    43
20302    45
Length: 25800, dtype: int64

In [0]:
dic= df3.to_dict()

In [0]:
contador = 0

for key, value in dic.items(): 
    if(dic[key]>=7):
        contador+=1
contador

5152

In [0]:
dic.items()

dict_items([(27685, 1), (19939, 1), (36005, 1), (19945, 1), (36009, 1), (8796, 1), (8795, 1), (36011, 1), (19968, 1), (8785, 1), (36022, 1), (8779, 1), (8776, 1), (8773, 1), (31021, 1), (8768, 1), (19975, 1), (8765, 1), (19977, 1), (19979, 1), (19984, 1), (36024, 1), (36025, 1), (19996, 1), (8747, 1), (31017, 1), (8812, 1), (36003, 1), (19931, 1), (19929, 1), (31046, 1), (8862, 1), (35986, 1), (8859, 1), (35987, 1), (8857, 1), (19900, 1), (19907, 1), (19909, 1), (8846, 1), (35993, 1), (19921, 1), (20003, 1), (8836, 1), (8833, 1), (36000, 1), (19923, 1), (31041, 1), (8827, 1), (8825, 1), (8824, 1), (8823, 1), (8822, 1), (19926, 1), (36002, 1), (31040, 1), (8835, 1), (35983, 1), (36035, 1), (20016, 1), (36061, 1), (20064, 1), (30982, 1), (36062, 1), (8668, 1), (20067, 1), (20068, 1), (20069, 1), (20078, 1), (8659, 1), (36070, 1), (8652, 1), (20081, 1), (20088, 1), (20089, 1), (20090, 1), (8643, 1), (20096, 1), (20101, 1), (8633, 1), (20102, 1), (8629, 1), (30966, 1), (8627, 1), (30964, 1

In [0]:
arregloIDsAEliminar = []

for key, value in dic.items(): 
    if(dic[key]>=7):
        arregloIDsAEliminar.append(key)
        
arregloIDsAEliminar
        

[16401,
 37782,
 26903,
 33927,
 25213,
 36109,
 5164,
 27363,
 31142,
 11929,
 16384,
 37625,
 19815,
 3583,
 13796,
 6863,
 11977,
 28002,
 4810,
 17219,
 11377,
 9562,
 11968,
 23397,
 38739,
 3841,
 16392,
 21777,
 19677,
 39373,
 37765,
 21706,
 37691,
 27798,
 16787,
 21893,
 9789,
 3760,
 13864,
 8333,
 23225,
 5049,
 32412,
 25323,
 36155,
 13686,
 3761,
 13832,
 31223,
 515,
 17062,
 4895,
 11755,
 11741,
 9687,
 34833,
 3727,
 21842,
 14011,
 1079,
 14008,
 629,
 13976,
 5123,
 25220,
 17095,
 3012,
 16191,
 20441,
 22490,
 13100,
 33098,
 30791,
 36597,
 20440,
 15293,
 7950,
 7477,
 9037,
 846,
 20384,
 22591,
 31387,
 4092,
 9127,
 34117,
 34082,
 39282,
 8971,
 34180,
 8843,
 8902,
 4211,
 22375,
 22379,
 20929,
 39158,
 24920,
 13036,
 28853,
 4406,
 8968,
 21015,
 28240,
 30749,
 39849,
 13073,
 16200,
 667,
 22622,
 38643,
 13283,
 38655,
 20050,
 36082,
 9387,
 8150,
 24577,
 26930,
 9401,
 16014,
 4705,
 2993,
 27866,
 37829,
 38226,
 16167,
 29085,
 26979,
 25109,
 

In [0]:
    df.drop(df.index[[arregloIDsAEliminar]])
    


Var6  Var7    Var13   Var21   Var22  Var24  Var25   Var28  Var35  \
0      1526.0   7.0    184.0   464.0   580.0   14.0  128.0  166.56    0.0   
1       525.0   0.0      0.0   168.0   210.0    2.0   24.0  353.52    0.0   
2      5236.0   7.0    904.0  1212.0  1515.0   26.0  816.0  220.08    0.0   
4      1029.0   7.0   3216.0    64.0    80.0    4.0   64.0  200.00    0.0   
5       658.0   7.0   3156.0   224.0   280.0    2.0   72.0  200.00    5.0   
6        77.0   0.0      0.0    32.0    40.0    2.0   16.0  230.56    0.0   
7      1176.0   7.0   2912.0   200.0   250.0    2.0   64.0  300.32    0.0   
8      1141.0   7.0    164.0   208.0   260.0    2.0   72.0  166.56    5.0   
9       490.0   7.0   2060.0    92.0   115.0    6.0  112.0  133.12    0.0   
10      798.0  14.0   2892.0   236.0   295.0    8.0   40.0  133.12    5.0   
11      595.0   0.0   2044.0     0.0     0.0    NaN    0.0  240.56    0.0   
12     2268.0   0.0      0.0   480.0   600.0   10.0  216.0  176.56    0.0   
13     3633.0   7.0    500.0   148.0   185.0    0.0    8.0  236.08    0.0   
14      259.0   0.0    348.0     8.0    10.0    0.0    0.0  166.56    0.0   
15     1449.0   7.0     28.0   168.0   210.0    2.0   32.0  166.56    5.0   
16      574.0   7.0   1752.0    12.0    15.0    2.0    0.0  253.52    0.0   
17      245.0   0.0      0.0   168.0   210.0    2.0   56.0  272.08    0.0   
18       14.0   0.0      0.0    20.0    25.0    2.0    0.0   86.96    0.0   
19      826.0   0.0     24.0   192.0   240.0    2.0   80.0  166.56    0.0   
20     1043.0   7.0    652.0   152.0   190.0    4.0   16.0   20.08   10.0   
21     1533.0   7.0     84.0   332.0   415.0    6.0  208.0  133.12    0.0   
22      924.0  28.0   4896.0   112.0   140.0    4.0   56.0  166.56    0.0   
23      469.0   7.0   4688.0    28.0    35.0    0.0   16.0  285.20    0.0   
24      833.0  14.0   7124.0   132.0   165.0    2.0   24.0  220.08    0.0   
25     1008.0  14.0   2616.0   160.0   200.0    4.0   40.0  133.12    0.0   
26     2198.0  14.0   5324.0   512.0   640.0   10.0  304.0  133.12    0.0   
28     1813.0   0.0      0.0   612.0   765.0   14.0  360.0  200.00    0.0   
29      931.0   7.0    108.0    76.0    95.0    0.0   16.0  213.36    0.0   
30     1484.0   7.0    120.0   208.0   260.0    4.0  112.0  203.44    0.0   
31      714.0   7.0   2240.0   120.0   150.0   10.0   80.0  133.12    0.0   
...       ...   ...      ...     ...     ...    ...    ...     ...    ...   
40015   196.0   0.0      0.0   160.0   200.0    4.0   72.0  280.56    0.0   
40016  1575.0   7.0    352.0   512.0   640.0    4.0  232.0  166.56    0.0   
40018  5726.0  14.0  12524.0   176.0   220.0    4.0   80.0  240.16    0.0   
40020   637.0   0.0      0.0    76.0    95.0    2.0   24.0  220.08    0.0   
40021  1337.0   7.0   1056.0   140.0   175.0    4.0   64.0  253.52    0.0   
40022   546.0   7.0    192.0   192.0   240.0    0.0   72.0  166.56    0.0   
40023  1659.0   0.0      0.0    16.0    20.0    0.0    0.0  220.08    0.0   
40024   133.0   7.0     76.0    36.0    45.0    0.0   16.0  200.00    0.0   
40026   742.0   7.0   2448.0   124.0   155.0    0.0   16.0  220.08   10.0   
40027  1218.0   7.0   3416.0   188.0   235.0    4.0   64.0  233.44    0.0   
40028   602.0   7.0   4952.0   176.0   220.0    4.0  128.0  238.80    0.0   
40029   154.0   0.0      0.0   136.0   170.0    4.0   16.0  253.52    0.0   
40031   714.0   7.0    304.0   388.0   485.0    2.0   72.0  262.00    0.0   
40032  1841.0   7.0   8460.0   288.0   360.0    4.0   56.0  510.88   10.0   
40033  6671.0   7.0   2220.0  1012.0  1265.0    4.0  608.0  553.84    0.0   
40034   511.0   7.0    104.0   116.0   145.0    4.0   96.0  167.20    0.0   
40035  1470.0  14.0   2048.0   300.0   375.0    2.0  192.0  166.56    0.0   
40037  1302.0   7.0   6040.0   128.0   160.0    0.0    8.0  186.64    0.0   
40038  6853.0   7.0      8.0  1268.0  1585.0   32.0  600.0  133.12    0.0   
40039   546.0   0.0      0.0   112.0   140.0    0.0    0.0  220.08    0.0   
40041